notebooks/02-ingestion/04-table-datatypes.ipynb

```diff
@@ -0,0 +1,851 @@
+    "  \"Reporting_Airline\",\n",
+    "  \"Origin\",\n",
+    "  CASE WHEN \"Origin\" = 'SFO' THEN true ELSE false END AS \"isSilicon\",\n",
+    "  \"Dest\"\n",
+    "FROM \"ext\"\n",
+    "WHERE \"arrivalime\" <> 0\n",
+    "AND \"depaturetime\" <> 0\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "\n",
+    "display.run_task(sql)\n",
+    "sql_client.wait_until_ready('example-flights-types-4')\n",
+    "display.table('example-flights-types-4')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3a4f215a-7f10-42b8-94f6-6037f161996f",
+   "metadata": {},
+   "source": [
+    "Notice the data types of `Distance` and of the three `speed` columns.\n",
+    "\n",
+    "* `Distance` has a data type of BIGINT.\n",
+    "* `doubleDistance` has a data type of DOUBLE.\n",
+    "* `speed` inherits the data type of `Distance`.\n",
+    "* `doubleSpeed` inherits the data type of `Distance` CAST to a DOUBLE.\n",
+    "* `floatSpeed` inherits the data type of `Distance` CAST to a FLOAT.\n",
+    "\n",
+    "Run the following query to see how the data differs between each one."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3397e916-d16c-43c2-aed9-ff9c4b151f6f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "SELECT\n",
+    "  \"Origin\",\n",
+    "  \"Dest\",\n",
+    "  \"Distance\",\n",
+    "  \"doubleDistance\",\n",
+    "  \"speed\",\n",
+    "  \"doubleSpeed\",\n",
+    "  \"floatSpeed\"\n",
+    "FROM \"example-flights-types-4\"\n",
+    "WHERE TIME_IN_INTERVAL(\"__time\",'2005-11-08/PT15M')\n",
+    "'''\n",
+    "\n",
+    "display.sql(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3e7fb5d5-c406-4df1-92d2-41fbe2ad9d9e",
+   "metadata": {},
+   "source": [
+    "## Strings\n",
+    "\n",
+    "CHAR and VARCHAR data is stored as a string, and can be understood as holding single or [multiple](https://druid.apache.org/docs/latest/querying/sql-data-types#multi-value-strings) values.\n",
+    "\n",
+    "> [Multi-value strings](https://druid.apache.org/docs/latest/querying/multi-value-dimensions) exhibit [special behaviour](https://druid.apache.org/docs/latest/querying/sql-data-types#multi-value-strings-behavior). A full exploration is out of the scope of this notebook."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ea07299d-27e2-407d-8df7-2bee1127d444",
+   "metadata": {},
+   "source": [
+    "In the table above, the `DayofMonth_Filter` column is recognized as a VARCHAR because of the CAST function.\n",
+    "\n",
+    "Since Druid automatically creates [bitmap indexes](https://druid.apache.org/docs/latest/ingestion/ingestion-spec#dimension-objects) for STRING-type columns at ingestion time, it can be useful to create VARCHAR versions of columns that will be used for filtering in addition to any CLUSTERED BY column, especially when that column will never be used for any sort of mathematical calculation.\n",
```

Comment:
"Since Druid automatically creates [bitmap indexes](https://druid.apache.org/docs/latest/ingestion/ingestion-spec#dimension-objects) for STRING-type columns at ingestion time, it can be useful to create VARCHAR versions of columns that will be used for filtering in addition to any CLUSTERED BY column, especially when that column will never be used for any sort of mathematical calculation."

This is no longer true when you ingest numeric columns using "auto" data type. It will create indexes on numeric columns as well.

Reply:
Maybe an R28 difference....I think this is in 27.... not sure

Reply:
Checking the latest version I can see today of https://github.com/apache/druid/blob/master/docs/ingestion/schema-design.md, numeric columns still don't have indexes. I did a hunt of PRs etc. and couldn't find anything. I'll take this as is for now (ie we have a 27 version) and then will update later.

Reply:
Actually I changed this notebook to just link out to that page instead - then it's up to the reader to go and look at what the current implementation details are :)